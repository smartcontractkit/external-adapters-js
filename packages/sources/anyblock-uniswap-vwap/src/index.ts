import { expose } from '@chainlink/ea-bootstrap'
import { execute } from './adapter'

const NAME = 'ANYBLOCK_UNISWAP_VWAP'

export = { NAME, execute, ...expose(execute) }
