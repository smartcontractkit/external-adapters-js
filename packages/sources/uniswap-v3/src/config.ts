import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as BaseConfig, ConfigFactory } from '@chainlink/types'
import { ethers } from 'ethers'

export const NAME = 'UNISWAP_V3'

export const ENV_ETHEREUM_RPC_URL = 'ETHEREUM_RPC_URL'
export const ENV_FALLBACK_RPC_URL = 'RPC_URL'
export const ENV_BLOCKCHAIN_NETWORK = 'BLOCKCHAIN_NETWORK'
export const ENV_QUOTER_CONTRACT = 'QUOTER_CONTRACT'
export const ENV_FEE_TIERS = 'DEFAULT_FEE_TIERS'

export const DEFAULT_ENDPOINT = 'crypto'
export const DEFAULT_BLOCKCHAIN_NETWORK = 'ethereum'

// https://docs.uniswap.org/protocol/concepts/V3-overview/fees#pool-fees-tiers
// LOW    MID    HIGH
// 0.05%  0.30%  1.00%
// 500    3000   10000
export const DEFAULT_FEE_TIERS = '500,3000,10000' //[500, 3_000, 10_000]
export const DEFAULT_QUOTER_CONTRACT = '0xb27308f9F90D607463bb33eA1BeBb41C27CE5AB6'

export type Config = BaseConfig & {
  provider: ethers.providers.Provider
  network: string
  uniswapQuoter: string
  feeTiers: number[]
}

export const makeConfig: ConfigFactory<Config> = (prefix: string | undefined) => {
  let parsedFeeTiers = DEFAULT_FEE_TIERS.split(',').map((f) => parseInt(f))
  if (util.getEnv(ENV_FEE_TIERS, prefix)) {
    parsedFeeTiers = util
      .getEnv(ENV_FEE_TIERS, prefix)!
      .split(',')
      .map((f) => parseInt(f))
  }

  return {
    ...Requester.getDefaultConfig(prefix),
    defaultEndpoint: DEFAULT_ENDPOINT,
    provider: new ethers.providers.JsonRpcProvider(
      util.getRequiredEnvWithFallback(ENV_ETHEREUM_RPC_URL, [ENV_FALLBACK_RPC_URL], prefix),
    ),
    network: util.getEnv(ENV_BLOCKCHAIN_NETWORK, prefix) || DEFAULT_BLOCKCHAIN_NETWORK,
    uniswapQuoter: util.getEnv(ENV_QUOTER_CONTRACT, prefix) || DEFAULT_QUOTER_CONTRACT,
    feeTiers: parsedFeeTiers,
  }
}
