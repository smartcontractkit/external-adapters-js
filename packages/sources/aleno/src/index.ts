import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
