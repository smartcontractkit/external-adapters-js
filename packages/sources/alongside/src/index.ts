import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { collateral } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: collateral.name,
  name: 'ALONGSIDE',
  config,
  endpoints: [collateral],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
