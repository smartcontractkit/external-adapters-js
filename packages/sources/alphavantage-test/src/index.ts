import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { PriceAdapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { forex } from './endpoint'

export const adapter = new PriceAdapter({
  defaultEndpoint: forex.name,
  name: 'ALPHAVANTAGE',
  config,
  endpoints: [forex],
  rateLimiting: {
    tiers: {
      free: {
        rateLimit1m: 5,
        rateLimit1h: 20.83,
      },
      '49.99': {
        rateLimit1m: 75,
      },
      '99.99': {
        rateLimit1m: 150,
      },
      '149.99': {
        rateLimit1m: 300,
      },
      '199.99': {
        rateLimit1m: 600,
      },
      '249.99': {
        rateLimit1m: 1200,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
