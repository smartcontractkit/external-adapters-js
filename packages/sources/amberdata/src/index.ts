import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig, WSHandlerFactory, NAME } from './config'

export = { NAME, makeExecute, makeConfig, ...expose(makeExecute(), WSHandlerFactory()) }
