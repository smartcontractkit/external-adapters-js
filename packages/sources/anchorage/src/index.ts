import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { config } from './config'
import { wallet } from './endpoint'
import { PoRAdapter } from '@chainlink/external-adapter-framework/adapter/por'

export const adapter = new PoRAdapter({
  defaultEndpoint: wallet.name,
  name: 'ANCHORAGE',
  config,
  endpoints: [wallet],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
        note: 'Docs: 10 requests per second per Organization, bursts of up to 100 requests per second',
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
export const walletParameters = wallet.inputParameters
