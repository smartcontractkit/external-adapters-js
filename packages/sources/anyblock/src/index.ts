import { expose } from '@chainlink/ea-bootstrap'
import { endpointSelector, makeExecute } from './adapter'
import { makeConfig } from './config'

const NAME = 'ANYBLOCK'

export = {
  NAME,
  makeExecute,
  makeConfig,
  ...expose(NAME, makeExecute(), undefined, endpointSelector),
}
