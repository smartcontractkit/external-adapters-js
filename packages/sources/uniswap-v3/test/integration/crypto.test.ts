import { AdapterRequest } from '@chainlink/types'
import request from 'supertest'
import process from 'process'
import nock from 'nock'
import http from 'http'
import { server as startServer } from '../../src'
import { mockEthereumResponseSuccess } from './fixtures'

let oldEnv: NodeJS.ProcessEnv

beforeAll(() => {
  oldEnv = JSON.parse(JSON.stringify(process.env))
  process.env.RPC_URL = process.env.RPC_URL || 'http://localhost:8545'
  process.env.API_VERBOSE = true
  if (process.env.RECORD) {
    nock.recorder.rec()
  }
})

afterAll(() => {
  process.env = oldEnv
  if (process.env.RECORD) {
    nock.recorder.play()
  }

  nock.restore()
  nock.cleanAll()
  nock.enableNetConnect()
})

describe('execute', () => {
  const id = '1'
  let server: http.Server
  const req = request('localhost:8080')
  beforeAll(async () => {
    server = await startServer()
  })
  afterAll((done) => {
    server.close(done)
  })

  describe('with from/to', () => {
    const data: AdapterRequest = {
      id,
      data: {
        from: 'USDC',
        to: 'USDT',
      },
    }

    it('should return success', async () => {
      mockEthereumResponseSuccess()

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('with custom params', () => {
    const data: AdapterRequest = {
      id,
      data: {
        from: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2', // USDC
        fromDecimals: 18,
        to: '0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984', // UNI
        toDecimals: 18,
        amount: 10,
      },
    }

    it('should return success', async () => {
      mockEthereumResponseSuccess()

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})
