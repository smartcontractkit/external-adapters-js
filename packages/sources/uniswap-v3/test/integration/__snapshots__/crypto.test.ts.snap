// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute with custom params should return success 1`] = `
Object {
  "data": Object {
    "input": "10000000000000000000",
    "inputDecimals": 18,
    "inputToken": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    "output": "1454105298115974341987",
    "outputDecimals": 18,
    "outputToken": "0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984",
    "rate": 145.41052981159743,
    "result": 145.41052981159743,
  },
  "jobRunID": "1",
  "result": 145.41052981159743,
  "statusCode": 200,
}
`;

exports[`execute with from/to should return success 1`] = `
Object {
  "data": Object {
    "input": "1000000",
    "inputDecimals": 6,
    "inputToken": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    "output": "999148",
    "outputDecimals": 6,
    "outputToken": "0xdAC17F958D2ee523a2206206994597C13D831ec7",
    "rate": 0.999148,
    "result": 0.999148,
  },
  "jobRunID": "1",
  "result": 0.999148,
  "statusCode": 200,
}
`;
